### PR TITLE
[BUG][Resource Center] Task can not select main JAR

### PR DESCRIPTION
**Pull request by Radeity**
## Purpose of the pull request

- Task can not select main jar (like Spark task), because `queryResourceByProgramType` have not been updated after refactoring the resource center.

## Brief change log

- Rewrite `ResourcesServiceImpl.queryResourceByProgramType` with right logic, which fit the new resource center.
- Update `ResourceFilter`, directly filter `StorageEntity`, rather deprecated `Resource`.
- Clean part of the code in `ResourcesServiceImpl`.
- Remove useless `res` in `ResourceInfo`.
- Fix some relative bug in front-end.
- Set field `resources` in `TaskInstance` and `TaskExecutionContext` to <`resourceFullName`, `resourceFileName`>.

## Verify this pull request

- Covered by existed UT.

### CODE REVIEW

**Review by SbloodyS — DISMISSED** [review]
LGTM. Thanks.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: Why not use `taskDefinition.getId()` directly?

Comment (Radeity) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: The type of taskDefinition here is `taskDefinitionLog`, however, resource binds with `taskDefinition`.

Comment (caishunfeng) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: > The type of taskDefinition here is `taskDefinitionLog`, however, resource binds with `taskDefinition`.

Get it. What about expand `updateResourceInfo` with taskCode and version? like 
`protected ResourceInfo updateResourceInfo(long taskCode, int taskVersion, ResourceInfo res)`,  and need to add some comments for it.

Comment (Radeity) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: > What about expand updateResourceInfo with taskCode and version?

May i ask why we have to add `taskVersion`?

Comment (caishunfeng) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: > May i ask why we have to add `taskVersion`?

Yes, I think we need to add code and version, `taskDefinitionMapper.queryByCode(taskDefinition.getCode())` means query the current using version, but current version maybe is not queal `taskDefinition.version`.

Comment (Radeity) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1375]: > but current version maybe is not equal taskDefinition.version.

Current version may be different, but they have the same code, which is equal to `taskDefinition.code`, so that we can query the right resource task relation.

Comment (Radeity) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:1381]: > but current version maybe is not equal taskDefinition.version.

Current version may be different, but they have the same code, which is equal to `taskDefinition.code`, so that we can query the right resource task relation.

**Comment by Radeity** [comment]
I just find another error, which confuse me. May i ask why we have `res` in `ResourceInfo`, can we directly use `resourceName` (`fullName`) to generate the submit command? @caishunfeng @SbloodyS 

```java
return mainJar.getId() == null
                ? mainJar.getRes()
                // when update resource maybe has error
                : mainJar.getResourceName().replaceFirst("/", "");
```
The above codes return wrong `resName`, for example, `resourceName` is `hdfs://xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`, after replacing, it will return `hdfs:/xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`.

**Comment by Radeity** [comment]
> I just find another error, which confuse me. May i ask why we have `res` in `ResourceInfo`, can we directly use `resourceName` (`fullName`) to generate the submit command? @caishunfeng @SbloodyS
> 
> ```java
> return mainJar.getId() == null
>                 ? mainJar.getRes()
>                 // when update resource maybe has error
>                 : mainJar.getResourceName().replaceFirst("/", "");
> ```
> 
> The above codes return wrong `resName`, for example, `resourceName` is `hdfs://xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`, after replacing, it will return `hdfs:/xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`.


I've checked and found that `res` is useless, remove relative read and write, PTAL @caishunfeng @SbloodyS  @EricGao888, thx!

**Review by EricGao888 — DISMISSED** [review]
Basically LGTM, would u like to double-check when available? Thx @SbloodyS @caishunfeng

**Comment by rickchengx** [comment]
> > I just find another error, which confuse me. May i ask why we have `res` in `ResourceInfo`, can we directly use `resourceName` (`fullName`) to generate the submit command? @caishunfeng @SbloodyS
> > ```java
> > return mainJar.getId() == null
> >                 ? mainJar.getRes()
> >                 // when update resource maybe has error
> >                 : mainJar.getResourceName().replaceFirst("/", "");
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The above codes return wrong `resName`, for example, `resourceName` is `hdfs://xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`, after replacing, it will return `hdfs:/xxx:9000/dolphinscheduler/resources/spark-examples_2.12-3.1.2.jar`.
> 
> I've checked and found that `res` is useless, remove relative read and write, PTAL @caishunfeng @SbloodyS @EricGao888, thx!

Hi, @Radeity , May I ask is `res` in `ResourceInfo` completely useless？If yes, is it possible to consider removing it, since it seems that `res` and `resourceName` are confused

**Comment by Radeity** [comment]
> Hi, @Radeity , May I ask is `res` in `ResourceInfo` completely useless？If yes, is it possible to consider removing it, since it seems that `res` and `resourceName` are confused

Hi, @rickchengx , you're right, it should be removed. I plan to do it in https://github.com/apache/dolphinscheduler/issues/13570. Like in method `getResourceFullNames`:
https://github.com/apache/dolphinscheduler/blob/c406f81e9c78cfda7344142b253b686f04ced3bc/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/StreamTaskExecuteRunnable.java#L383-L386

`res` is still in use, however, `getResourceFullNames` is useless after refactoring the resource center, i'll clean these codes in later pr.

**Comment by Radeity** [comment]
Hi, @SbloodyS , @caishunfeng , @EricGao888 , @rickchengx , I've made some changes that submit task with local path of downloaded resource in `executePath`, rather directly with `fullName` in resource center, please help review it again, thx!

**MORE EXPLANATION**: DS should be responsible for resource download, if submit with `fullName`, such as using `OSS` as resource center and run a `Spark` task, user may have to add `OSS` plugin in `Spark` classpath additionally.

**Comment by Radeity** [comment]
Hi, @caishunfeng @SbloodyS @EricGao888 @ruanwenjun , i think this pr has a relatively high priority, would you like to help review it when you're available, thx!

**Review by caishunfeng — DISMISSED** [review]
Basically LGTM

**Comment by caishunfeng** [comment]
Hi @Radeity please resolve the conflicts

**Comment by Radeity** [comment]
> Hi @Radeity please resolve the conflicts

I'll handle that later, thx!

**Comment by Radeity** [comment]
> Hi @Radeity please resolve the conflicts

Done.

**Review by SbloodyS — APPROVED** [review]
LGTM

**Comment by Yinyuhang123** [comment]
> > Hi @Radeity please resolve the conflicts
> 
> Done.
May I ask which branch did you submit this commit to? Why didn't I find it in 3.1.4, 3.1.5 or dev

**Comment by Yinyuhang123** [comment]
> > Hi @Radeity please resolve the conflicts
> 
> Done.

May I ask which branch did you submit this commit to? Why didn't I find it in 3.1.4, 3.1.5 or dev

**Comment by Radeity** [comment]
> > > Hi @Radeity please resolve the conflicts
> > 
> > 
> > Done.
> 
> May I ask which branch did you submit this commit to? Why didn't I find it in 3.1.4, 3.1.5 or dev

Hi, @Yinyuhang123 , you can find it in branch dev after this pr is merged, released versions don't have this bug.

**Comment by Radeity** [comment]
Hi, @caishunfeng, would you like to help double check?

**Review by SbloodyS — COMMENTED** [review]
I'll merge this PR in the afternoon if there are no other comments.

**Pull request merged**
